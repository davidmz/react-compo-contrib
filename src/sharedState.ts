import {
  StateSetter,
  StateSetterArg,
  Use,
  HookCreator,
  state,
  effector,
  SimpleEvents,
} from 'react-compo';

export type SharedStateUpdate<T> = StateSetter<T>;

interface SharedStateSubscriber<T> {
  <O>(selector: (v: T) => O): HookCreator<() => O>;
  (): HookCreator<() => T>;
}

const UPDATE = 'UPDATE';
const identity = <T>(x: T) => x;

export function createSharedState<T>(
  initial: T
): [SharedStateSubscriber<T>, SharedStateUpdate<T>] {
  const events = new SimpleEvents();
  let val = initial;

  return [
    // Subscribe
    (selector: (x: any) => any = identity) => (use: Use) => {
      const [, set] = use(state(selector(val)));
      const onUpdate = () => set(selector(val));

      const effect = use(effector());
      effect(() => {
        events.on(UPDATE, onUpdate);
        return () => events.off(UPDATE, onUpdate);
      }, []);

      return () => selector(val);
    },

    // Update
    (v: StateSetterArg<T>) => {
      if (typeof v === 'function') {
        v = (v as ((v: T) => T))(val);
      }
      if (val !== v) {
        val = v;
        events.emit(UPDATE);
      }
    },
  ];
}
