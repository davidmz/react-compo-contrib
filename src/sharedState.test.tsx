import React, { ReactNode } from 'react';
import { mount } from 'enzyme';
import { compo, useWith, DO_UPDATE, DID_UPDATE } from 'react-compo';

import { createSharedState } from './sharedState';
import { EventEmitter } from 'events';

describe('sharedState', async () => {
  it('should propagate the shared state', () => {
    const [subscribe, update] = createSharedState(42);

    const Leaf = compo('Leaf', (use) => {
      const getNum = use(subscribe());
      return () => <p>The number is {getNum()}</p>;
    });

    function Tree() {
      return (
        <div>
          <Leaf />
          <Leaf />
          <button onClick={() => update((n) => n + 1)}>+1</button>
        </div>
      );
    }

    const wrapper = mount(<Tree />);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('button').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('should propagate shared state with selector', () => {
    const [subscribe, update] = createSharedState({ a: 1, b: 2 });

    const Leaf = compo('Leaf', (use) => {
      const getA = use(subscribe((s) => s.a));
      return () => <p>The 'a' field is {getA()}</p>;
    });

    const Tree = compo('Tree', () => {
      return () => (
        <div>
          <Leaf />
          <Leaf />
          <button onClick={() => update((s) => ({ ...s, a: s.a + 1 }))}>
            a + 1
          </button>
        </div>
      );
    });

    const wrapper = mount(<Tree />);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('button').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('should not update component if selected value is not changed', () => {
    const [subscribe, update] = createSharedState({ a: 1, b: 2 });

    const leafRender = jest.fn();

    const Leaf = compo('Leaf', (use) => {
      const getA = use(subscribe((s) => s.a));
      return () => (
        <p>
          {leafRender()}The 'a' field is {getA()}
        </p>
      );
    });

    const Tree = compo('Tree', () => {
      return () => (
        <div>
          <Leaf />
          <button onClick={() => update((s) => ({ ...s, b: s.b + 1 }))}>
            b + 1
          </button>
        </div>
      );
    });

    const wrapper = mount(<Tree />);
    leafRender.mockClear();
    wrapper.find('button').simulate('click');
    expect(leafRender).not.toBeCalled();
  });

  it('should not update nested components multiple times', () => {
    const [subscribe, update] = createSharedState(42);

    const mocks: jest.Mock<{}>[] = [jest.fn(), jest.fn(), jest.fn(), jest.fn()];

    const Leaf = compo('Leaf', (use) => {
      use(subscribe()); // just subscribe to state
      return ({ idx, children }: { idx: number; children: ReactNode }) => {
        mocks[idx]();
        return children;
      };
    });

    const Tree = () => (
      <div>
        <Leaf idx={0}>
          <Leaf idx={1}>
            <Leaf idx={3}>end</Leaf>
          </Leaf>
          <Leaf idx={2}>end</Leaf>
        </Leaf>
        <button onClick={() => update((n) => n + 1)}>+1</button>
      </div>
    );

    const wrapper = mount(<Tree />);
    jest.clearAllMocks();
    wrapper.find('button').simulate('click');

    mocks.forEach((m) => expect(m).toBeCalledTimes(1));
  });

  describe('Without React', () => {
    it('should propagate the shared state', () => {
      const e1 = new EventEmitter();
      const e2 = new EventEmitter();
      const upd1 = jest.fn();
      const upd2 = jest.fn();
      e1.on(DO_UPDATE, upd1);
      e2.on(DO_UPDATE, upd2);

      const [subscribe, update] = createSharedState(42);
      const get1 = useWith(e1)(subscribe());
      const get2 = useWith(e2)(subscribe());

      // After the "first render"
      e1.emit(DID_UPDATE);
      e2.emit(DID_UPDATE);

      expect(upd1).not.toBeCalled();
      expect(upd2).not.toBeCalled();

      expect(get1()).toBe(42);
      expect(get2()).toBe(42);

      update(43);

      expect(upd1).toBeCalled();
      expect(upd2).toBeCalled();

      expect(get1()).toBe(43);
      expect(get2()).toBe(43);
    });

    it('should propagate state with selector', () => {
      const e = new EventEmitter();
      const upd = jest.fn();
      e.on(DO_UPDATE, upd);

      const [subscribe, update] = createSharedState(42);
      const get = useWith(e)(subscribe((n) => n % 2));

      // After the "first render"
      e.emit(DID_UPDATE);

      expect(upd).not.toBeCalled();
      expect(get()).toBe(0);

      update(43);

      expect(upd).toBeCalled();
      expect(get()).toBe(1);

      upd.mockClear();
      update(45); // should not change selected value
      expect(upd).not.toBeCalled();
      expect(get()).toBe(1);
    });
  });
});
