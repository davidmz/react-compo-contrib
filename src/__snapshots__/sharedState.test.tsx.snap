// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sharedState should propagate shared state with selector 1`] = `
<Tree>
  <div>
    <Leaf>
      <p>
        The 'a' field is 
        1
      </p>
    </Leaf>
    <Leaf>
      <p>
        The 'a' field is 
        1
      </p>
    </Leaf>
    <button
      onClick={[Function]}
    >
      a + 1
    </button>
  </div>
</Tree>
`;

exports[`sharedState should propagate shared state with selector 2`] = `
<Tree>
  <div>
    <Leaf>
      <p>
        The 'a' field is 
        2
      </p>
    </Leaf>
    <Leaf>
      <p>
        The 'a' field is 
        2
      </p>
    </Leaf>
    <button
      onClick={[Function]}
    >
      a + 1
    </button>
  </div>
</Tree>
`;

exports[`sharedState should propagate the shared state 1`] = `
<Tree>
  <div>
    <Leaf>
      <p>
        The number is 
        42
      </p>
    </Leaf>
    <Leaf>
      <p>
        The number is 
        42
      </p>
    </Leaf>
    <button
      onClick={[Function]}
    >
      +1
    </button>
  </div>
</Tree>
`;

exports[`sharedState should propagate the shared state 2`] = `
<Tree>
  <div>
    <Leaf>
      <p>
        The number is 
        43
      </p>
    </Leaf>
    <Leaf>
      <p>
        The number is 
        43
      </p>
    </Leaf>
    <button
      onClick={[Function]}
    >
      +1
    </button>
  </div>
</Tree>
`;
